fix: follow standard pattern for auto-generated site names

tinfoil-knight opened the pull request:
🎉 Thanks for submitting a pull request! 🎉

#### Summary

Fixes #4630
Fixes #4653

<!--
Explain the **motivation** for making this change. What existing problem does the pull request solve and how?
-->

The suggestions for auto-generated site names from the CLI don't follow the same pattern as they do when created from the Netlify App. These suggestions also use a very limited number of combinations which results in a user having a lot of site names being very similar & not being able to discern b/w them.

Additionally, the suggested site names can lead to invalid URLs as described here: https://github.com/netlify/cli/issues/4653#issue-1252491139

This PR fixes both the issue with standard names & site length. Also see https://github.com/netlify/cli/pull/5007#issuecomment-1235208464

---

For us to review and ship your PR efficiently, please perform the following steps:

- [x] Open a [bug/issue](https://github.com/netlify/cli/issues/new/choose) before writing your code 🧑‍💻. This ensures we can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
- [x] Read the [contribution guidelines](../CONTRIBUTING.md) 📖. This ensures your code follows our style guide and
      passes our tests.
- [ ] Update or add tests (if any source code was changed or added) 🧪
- [ ] Update or add documentation (if features were changed or added) 📝
- [x] Make sure the status checks below are successful ✅

**A picture of a cute animal (not mandatory, but encouraged)**

tinfoil-knight commented on src/utils/sites/utils.js:
```diff
@@ -55,4 +57,506 @@ const createRepo = async (templateName, ghToken, siteName) => {
 }
 
-module.exports = { getTemplatesFromGitHub, createRepo, validateTemplate }
+const SITE_NAME_SUGGESTION_SUFFIX_LENGTH = 6
```
I'm not sure of the logic behind this length. I've set it to 6 because all the sites created through the Netlify App UI had suffixes of this length.

Nonetheless, this length is valid.

tinfoil-knight commented:
The longest possible site name suggestion is `extraordinary-croquembouche-${suffix}` where suffix is of length 6 making the maximum length 34.

For deploy previous, a 24 char hash is pre-pended (as per https://github.com/netlify/cli/issues/4653#issue-1252491139) which would make the max length 58.

In all cases, the suggested URLs stay below the 63 char length.

danez commented:
I have mixed feelings about this. I do not like the code duplication from the server-side. In case we ever decide to change the logic server-side this can create problems again.

What if instead of creating a name on the client we simply do not do this at all? If the API call `createSiteInTeam` is called without `name` then the server generates a name. This makes this way more simple and we do not have any code duplication. I quickly tried this by removing all autogeneration and it seems to work fine.

Do you see any problem with this?

![Kapture 2022-08-29 at 14 09 31](https://user-images.githubusercontent.com/231804/187198100-794aa193-74e4-40c3-84e9-b48a368645ae.gif)

tinfoil-knight commented:
> What if instead of creating a name on the client we simply do not do this at all? If the API call `createSiteInTeam` is called without `name` then the server generates a name. This makes this way more simple and we do not have any code duplication. I quickly tried this by removing all autogeneration and it seems to work fine.
> 
> Do you see any problem with this?

Only issue with this is that it seems like the site name is going to be empty & there's no indication of what it's going to be.

Although, the Netlify UI also does the same thing (assign name without pre-indication).

I'll make the changes if you feel its necessary.

danez commented:
I personally feel this is okay if we add a message like `leave empty for autogenerated name`. 

Ping @netlify/developer-experience-members What is you take on this?

whitep4nth3r commented:
Something like `Press enter to {generate a random site name}` where 'Press enter' is the action, rather than 'Leave empty' which is an anti-action (kind of).

I left `generate a random site name` in curly brackets because I'm not sold on it.

tinfoil-knight commented:
@whitep4nth3r

How about `Skip to let Netlify pick a name for your site` ?

The prompt in the CLI would then be:

`? Site name (you can change it later): Skip to let Netlify pick a name for your site`

---

I don't want to go with "Press enter" because the key is named `return` for MacOS.

whitep4nth3r commented:
But how does someone `Skip`? That's not clear. I know we're getting deep into semantics and copy-writing and I'm not sure that's my area of expertise... but I'm not sure `Skip` instructs the user to press enter/return.

tinfoil-knight commented:
> But how does someone `Skip`? That's not clear. I know we're getting deep into semantics and copy-writing and I'm not sure that's my area of expertise... but I'm not sure `Skip` instructs the user to press enter/return.

I didn't want to use `Press enter` because it seems like an action step without much choice. `Skip` seems like it offers a choice to let the user not think of the site name (which is the intended use of auto-generated names).

I do agree that `Skip` might be a bit unclear but pressing enter or return to skip a step is usual behaviour for many CLI utilities (from anecdotal experience) so I don't think this should be too confusing.

whitep4nth3r commented:
You know, I'm not sure we need a message if the auto-generated name is a suggestion like in the image above from @tinfoil-knight.

I just ran through setting up a new git repo via the GitHub CLI for inspiration. All (lighter contrast) prompts are e.g. `(Y/n)` or`(.origin)` where the default option is chosen when you press enter  _without being told_. 

Maybe we just need to show the auto generated name without an extra message. That feels more intuitive to me. What do you think?

![image](https://user-images.githubusercontent.com/52798353/187412519-398c7c0c-1e00-4dc8-a8b8-f0220f3d9332.png)

tinfoil-knight commented:
> Maybe we just need to show the auto generated name without an extra message. That feels more intuitive to me. What do you think?

@whitep4nth3r I like the auto-generated name option from a user's PoV but the issue with this is that it uses code that relies on internal implementation (Netlify API) of how the site names are generated which makes it brittle which is why @danez suggested to avoid doing this in https://github.com/netlify/cli/pull/5007#issuecomment-1230202734

whitep4nth3r commented:
Oh I see, that makes sense.

So the options are along the lines of:

1. `Site name (you can change it later): leave empty for autogenerated name`
1. `Site name (you can change it later): skip to let Netlify pick a name for your site`

I wonder whether we can get a bit of Netlify personality in here. Maybe I'm over-thinking it but there's maybe an opportunity to be a little quirky here given we can't show the auto-generated name. 

Along the lines of....?

1.  `Site name (you can change it later): pick for me`
1.  `Site name (you can change it later): I'm feeling lucky`

Maybe we have an array of quirky strings to show a random one out of a selection each time? Am I going too far? 😅

tinfoil-knight commented:
`Site name (you can change it later): pick for me` sounds okay. I'd avoid going with an array of strings for consistency on the command line. 

@danez I'll go forward with this & make changes to the flow to rely on the server for generating names if left empty.

@whitep4nth3r If you come up with any other ideas in the meantime, just let me know.

tinfoil-knight commented:
Change Summary

`sites:create` will no longer auto-suggest names & will just a `Pick for me` text. If left as it is, name will be selected by the server.

<img width="474" alt="image" src="https://user-images.githubusercontent.com/51631122/188094350-f01f2eaf-754b-4338-b8f6-ec2b0e3c3b7a.png">

---

The `sites:create-template` command will now use the name of the template if the site name isn't available to create the Github Repository.

If a repository with the same name is already present, the user will get a well defined error.

The site name will be selected by the server if not provided.

The 
<img width="813" alt="image" src="https://user-images.githubusercontent.com/51631122/188094488-468d8cce-e936-46b3-a9c5-1427ec73c104.png">

tinfoil-knight reviewed (commented):
@danez Let me know if any additional tests are required.

eduardoboucas commented on src/commands/sites/sites-create.js:
```diff
@@ -1,58 +1,31 @@
         name: 'name',
         message: 'Site name (you can change it later):',
-        default: siteSuggestion,
+        default: 'Pick for me',
```
I don't think we should be using the `default` property like this. We're assigning it a value that we know we won't use, just for the purpose of displaying something.

Instead, can we change `message` so that it reads something like `Site name (leave blank for a random name; you can change it later)`, and we assign a value if the value is empty?

If we absolutely must use the current approach, let's please move the string to a constant.

tinfoil-knight replied:
I'll change it. I find the current approach a bit "hacky" too.

tinfoil-knight commented:
Not sure as to what's happening with the failing tests here: https://github.com/netlify/cli/actions/runs/2993351770

danez commented:
I'm not working till end of September. So i can look then or someone else does. Sorry

tinfoil-knight commented:
@danez Can you take a look at the tests here? They just seem to be timing out.

danez reviewed (approved):
LGTM!

tinfoil-knight commented:
> @tinfoil-knight I rebased on the current main branch and fixed the tests. It seems the tests match on the exact input line.

Thank you a lot for fixing the tests @danez.

verythorough commented:
This is awesome! Here I was going to post a list of the words we use to generate the site names in Netlify, and I see y'all settled on a better solution. ✨
